sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamSqlEnv.java

```diff
@@ -73,109 +80,182 @@ public static BeamSqlEnv inMemory(TableProvider... tableProviders) {
+    }
+
+    /**
+     * Register a UDAF function which can be used in GROUP-BY expression. See {@link
+     * org.apache.beam.sdk.transforms.Combine.CombineFn} on how to implement a UDAF.
+     */
+    public BeamSqlEnvBuilder registerUdaf(String functionName, Combine.CombineFn combineFn) {
+      functionSet.add(new SimpleEntry<>(functionName, new UdafImpl(combineFn)));
+      return this;
+    }
+
+    /** Load all UDF/UDAF from {@link UdfUdafProvider}. */
+    public BeamSqlEnvBuilder loadUdfUdafFromProvider() {
+      ServiceLoader.<UdfUdafProvider>load(UdfUdafProvider.class)
+          .forEach(
+              ins -> {
+                ins.getBeamSqlUdfs().forEach((udfName, udfClass) -> registerUdf(udfName, udfClass));
+                ins.getSerializableFunctionUdfs()
+                    .forEach((udfName, udfFn) -> registerUdf(udfName, udfFn));
+                ins.getUdafs().forEach((udafName, udafFn) -> registerUdaf(udafName, udafFn));
+              });
+
+      return this;
+    }
+
+    public BeamSqlEnvBuilder loadBeamBuiltinFunctions() {
+      for (BeamBuiltinFunctionProvider provider :
+          ServiceLoader.load(BeamBuiltinFunctionProvider.class)) {
+        registerBuiltinUdf(provider.getBuiltinMethods());
+      }
+
+      return this;
+    }
+
+    public BeamSqlEnvBuilder setQueryPlannerClassName(String name) {
+      queryPlannerClassName = name;
+      return this;
+    }
+
+    /**
+     * Build function to create an instance of BeamSqlEnv based on preset fields.
+     *
+     * @return BeamSqlEnv.
+     */
+    public BeamSqlEnv build() {
+      // This check is to retain backward compatible because most of BeamSqlEnv are initialized by
+      // withTableProvider API.
+      if (initialTableProvider == null) {
+        throw new RuntimeException("initialTableProvider must be set in BeamSqlEnvBuilder.");
+      }
+
+      JdbcConnection jdbcConnection = JdbcDriver.connect(initialTableProvider);
+
+      // set schema
+      for (Map.Entry<String, TableProvider> schemaEntry : schemaMap.entrySet()) {
+        jdbcConnection.setSchema(schemaEntry.getKey(), schemaEntry.getValue());
+      }
+
+      // reset default schema
+      if (currentSchemaName != null) {
+        try {
+          jdbcConnection.setSchema(currentSchemaName);
+        } catch (SQLException e) {
+          throw new RuntimeException(e);
+        }
+      }
+
+      // add UDF
+      for (Map.Entry<String, Function> functionEntry : functionSet) {
+        jdbcConnection.getCurrentSchemaPlus().add(functionEntry.getKey(), functionEntry.getValue());
+      }
+
+      QueryPlanner planner;
+
+      if (queryPlannerClassName.equals(
+          "org.apache.beam.sdk.extensions.sql.impl.CalciteQueryPlanner")) {
+        planner = new CalciteQueryPlanner(jdbcConnection);
+      } else {
+        try {
+          planner =
+              (QueryPlanner)
+                  Class.forName(queryPlannerClassName)
```

Comment:
The other way you can do this is with service loader. You can have a dictionary from `queryPlannerClassName` to `Class<? extends QueryPlanner>` and by AutoService insert things into the dictionary. You still use reflection to do `getConstructor` etc.